src/core-sdk-tasks/DownloadFile.cs

```diff
@@ -39,31 +54,70 @@ public override bool Execute()
                 var filePath = Uri.Substring(FileUriProtocol.Length);
                 Log.LogMessage($"Copying '{filePath}' to '{DestinationPath}'");
                 File.Copy(filePath, DestinationPath);
+                return true;
             }
-            else
+
+            var lastStatus = await DownloadWithRetries(Uri, DestinationPath);
+            if (lastStatus == false)
             {
-                Log.LogMessage(MessageImportance.High, $"Downloading '{Uri}' to '{DestinationPath}'");
+                if (!string.IsNullOrEmpty(PrivateUri))
+                {
+                    lastStatus = await DownloadWithRetries(PrivateUri, DestinationPath);
+                }
+            }
+
+            return lastStatus == true;
+        }
 
+        /// <summary>
+        /// Attempt to download file from `source` with retries when response error is different of FileNotFound and Success.
+        /// </summary>
+        /// <param name="source">URL to the file to be downloaded.</param>
+        /// <param name="target">Local path where to put the downloaded file.</param>
+        /// <returns>true: Download Succeeded. false: Download failed with 404. null: Download failed but is retriable.</returns>
+        private async Task<bool?> DownloadWithRetries(string source, string target)
+        {
+            Random rng = new Random();
+
+            Log.LogMessage(MessageImportance.High, $"Attempting download '{source}' to '{target}'");
+
+            for (int retryNumber = 0; retryNumber < MaxRetries; retryNumber++)
+            {
                 using (var httpClient = new HttpClient())
```

Comment:
It's actually better to use a single instance rather than create and dispose a bunch of `HttpClient`s. This doesn't seem like an extreme case but FYI. https://softwareengineering.stackexchange.com/q/330364